File: boj/2304/main.cpp
#include <iostream>
using namespace std;

int n, x, y, _max, ans = 0;
int arr[3][1001];

int main() {
  cin.tie(nullptr);
  ios_base::sync_with_stdio(false);

  cin >> n;
  for (int i = 0; i < n; i++) {
    cin >> x >> y;
    arr[0][x] = y;
  }

  for (int i = 0; i < 1001; i++) {
    _max = max(arr[0][i], _max);
    arr[1][i] = _max;
  }
  _max = 0;
  for (int i = 1000; i >= 0; i--) {
    _max = max(arr[0][i], _max);
    arr[2][i] = _max;
  }
  for (int i = 0; i < 1001; i++) {
    ans += min(arr[1][i], arr[2][i]);
  }
  cout << ans << '\n';
  return 0;
}

File: boj/2304/main.rs
/// ./main.cpp과 같은 로직으로 작성되었지만 RTE를 받음
use std::io::{self, BufRead};
use std::cmp::{min, max};

fn main() {
    let stdin = io::stdin();
    let pairs = stdin
        .lock()
        .lines()
        .skip(1)
        .map(|l| {
            l.unwrap()
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect::<Vec<i32>>()
        })
        .collect::<Vec<Vec<i32>>>();
    let mut arr = [0; 1002];
    let mut arr2 = [0; 1002];
    let mut arr3 = [0; 1002];

    for pair in pairs {
        let x = pair[0];
        let h = pair[1];
        arr[x as usize] = h;
    }

    let mut ans = 0;
    let mut _max = 0;

    for i in 0..1001 {
        _max = max(arr[i], _max);
        arr2[i] = _max;
    }
    _max = 0;
    for i in (0..1001).rev() {
        _max = max(arr[i], _max);
        arr3[i] = _max;
    }
    for i in 0..1001 {
        ans += min(arr2[i], arr3[i]);
    }
    println!("{}", ans);
}
